Feat: Add freshness + rows inserted graphs on the UI side#9117

ShaileshParmar11 opened the pull request:
### Describe your changes :
<!-- Explain what you have done & tag your assigned issue !-->
[Add freshness + rows inserted graphs on the UI side#9117](https://github.com/open-metadata/OpenMetadata/issues/9117)
Closes #9117 

#
### Type of change :
<!-- You should choose 1 option and delete options that aren't relevant -->
- [x] New feature

#
### Frontend Preview (Screenshots) :


https://user-images.githubusercontent.com/71748675/207606302-c3589aec-0358-4ff9-91d8-472461763cb7.mov




#
### Checklist:
<!-- add an x in [] if done, don't mark items that you didn't do !-->
- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#
### Reviewers
<!-- Please see the contributing guidelines and then add your reviewer(s) !-->
<!--- OpenMetadata community thanks you for explaining your changes in detail !-->
<!--- If you are unsure of people to review your work, you can add anyone of these developers :) !-->
<!--- Frontend: @open-metadata/ui -->
<!--- Backend: @open-metadata/backend -->
<!--- Ingestion: @open-metadata/ingestion -->

Sachin-chaurasiya commented on openmetadata-ui/src/main/resources/ui/src/components/Chart/CustomBarChart.tsx:
```diff
@@ -0,0 +1,104 @@
+}: CustomBarChartProps) => {
+  const { data, information } = chartCollection;
+  const { t } = useTranslation();
+  const renderColorfulLegendText = (
+    value: string,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    entry: any
```
I would recommend using `LegendProps['formatter']` to avoid explicit any type.

Sachin-chaurasiya commented on openmetadata-ui/src/main/resources/ui/src/components/Chart/CustomBarChart.tsx:
```diff
@@ -0,0 +1,104 @@
+    entry: any
+  ) => <span style={{ color: entry?.color }}>{value}</span>;
+
+  const tooltipFormatter = (value: string | number | (string | number)[]) => {
+    const numValue = value as number;
+
+    return (
+      <>
+        {tickFormatter
+          ? `${numValue.toFixed(2)}${tickFormatter}`
+          : formatNumberWithComma(numValue)}
+      </>
+    );
+  };
```
Same here we can use ` TooltipProps['formatter']`.

Sachin-chaurasiya commented on openmetadata-ui/src/main/resources/ui/src/components/Chart/CustomBarChart.tsx:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  return (
+    <ResponsiveContainer debounce={200} id={`${name}_graph`} minHeight={300}>
```
Why we are adding debounce as a higher `200`?

ShaileshParmar11 replied:
its 200 ms, so to creat graph we need to do some computation and it will take time so we need debounce

ShaileshParmar11 commented on openmetadata-ui/src/main/resources/ui/src/components/Chart/CustomBarChart.test.tsx:
```diff
@@ -0,0 +1,80 @@
+  name: 'testChart',
+};
+
+jest.mock('recharts', () => ({
+  Bar: jest.fn().mockImplementation(() => <div>Bar</div>),
+  Legend: jest.fn().mockImplementation(() => <div>Legend</div>),
+  Tooltip: jest.fn().mockImplementation(() => <div>Tooltip</div>),
+  XAxis: jest.fn().mockImplementation(() => <div>XAxis</div>),
+  YAxis: jest.fn().mockImplementation(() => <div>YAxis</div>),
+  BarChart: jest
+    .fn()
+    .mockImplementation(({ children }) => <div>{children}</div>),
+  ResponsiveContainer: jest
+    .fn()
+    .mockImplementation(({ children }) => (
+      <div data-testid="responsive-container">{children}</div>
+    )),
+}));
```
Yes library only provides container elements without child so we won't able to test without mock!

ShaileshParmar11 commented on openmetadata-ui/src/main/resources/ui/src/components/DatasetDetails/datasetDetails.style.less:
```diff
@@ -0,0 +1,21 @@
+  background: @white;
+  flex-grow: 1;
+  padding: 1rem /* 16px */;
+  box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06);
```
we need this style for lots of places, so we have to creat class at one place and use everywhere

ShaileshParmar11 commented on openmetadata-ui/src/main/resources/ui/src/components/TableProfiler/Component/TableProfilerChart.test.tsx:
```diff
@@ -0,0 +1,144 @@
+    .fn()
+    .mockImplementation(() => mockTimeValue.startMilli),
+}));
+jest.mock('../../../axiosAPIs/tableAPI');
```
Yes, and necessary methods are used in testing directly wherever it is needed!

ShaileshParmar11 commented on openmetadata-ui/src/main/resources/ui/src/components/TableProfiler/Component/TableProfilerChart.tsx:
```diff
@@ -0,0 +1,159 @@
+        startTs,
+        endTs,
+      });
+      const rowMetricsData = calculateRowCountMetrics(data, rowCountMetrics);
+      setRowCountMetrics(rowMetricsData);
```
Why do that? it will only increase function calls and additional storage for data, and delay in page load time!

Sachin-chaurasiya replied:
@ShaileshParmar11, I was recommending that change to just separate out the fetching logic and state data calculation.

```js
const rowMetricsData = calculateRowCountMetrics(data, rowCountMetrics);
      setRowCountMetrics(rowMetricsData);
```

So, If we want to use the above logic somewhere else then we can directly use that separate function.

Sachin-chaurasiya replied:
It was just a recommendation and not the required change, so we can go ahead without that.

ShaileshParmar11 commented on openmetadata-ui/src/main/resources/ui/src/axiosAPIs/tableAPI.ts:
```diff
@@ -179,6 +177,23 @@ export const getTableProfilesList = async (
+  const url = `/tables/${tableFqn}/systemProfile`;
+
+  const response = await APIClient.get<{
+    data: SystemProfile[];
+    paging: Paging;
```
done

ShaileshParmar11 commented:
Hi @Sachin-chaurasiya I have addressed the comments that made sense to me and marked it resolved! for others, I am open to discus!

harshach commented:
@ShaileshParmar11 on the side bar we kept as "profiler" which will be confusing as we are only displaying table level metrics cc @TeddyCr 
next thing is the fonts on the dates are huge in the profile charts

TeddyCr commented:
@ShaileshParmar11 looks amazing thanks. In addition to @harshach's comment can we make sure the colors are as follow (insert = green, update = blue, delete = purple)?

ShaileshParmar11 commented:
Hi @harshach @TeddyCr, I have updated the colours as suggested by @TeddyCr and changed the tab name to "Table Profiler" 
about font size, we are using the same as other chart details summary we can reduce it if needed cc @devyani-kaushik

harshach commented:
@TeddyCr is the bar graph really useful, can't seem to tell whats the chart is trying to tell the user

TeddyCr commented:
> @TeddyCr is the bar graph really useful, can't seem to tell whats the chart is trying to tell the user

@harshach mocks does not show a bar graph but a dot plot with thin bars linking to the timestamp (this is based on what I can see is done in the industry to display this type of info). @ShaileshParmar11 maube there is some technical limitation to display it this way?

ShaileshParmar11 commented:
> > @TeddyCr is the bar graph really useful, can't seem to tell whats the chart is trying to tell the user
> 
> @harshach wireframes don't show a bar graph but a dot plot with thin bars linking to the timestamp (this is based on what I can see is done in the industry to display this type of info). @ShaileshParmar11 maybe there is some technical limitation to display it this way?
> 
> ![Screenshot 2022-12-13 at 12 25 46 PM](https://user-images.githubusercontent.com/13626425/207305807-91ffc38a-6651-4f6d-8271-ed0c49bd06a5.png)
> 
> #9117

Yes, correct the chart is called the lollipop chart and the library does not have support for that type of chart!

TeddyCr commented:
> Yes, correct the chart is called the lollipop chart and the library does not have support for that type of chart! Can we go with Scatter chart? @TeddyCr

Sounds good to me based on the graphs available.

harshach commented:
@ShaileshParmar11 lets figure out library that support the chart in the mock

TeddyCr commented:
> @ShaileshParmar11 lets figure out library that support the chart in the mock

d3 seems to support it if we want to go that route. https://d3-graph-gallery.com/index.html

TeddyCr commented:
> 

@ShaileshParmar11 for this release we can go ahead with the scatter plot for the freshness. I opened a ticket https://github.com/open-metadata/OpenMetadata/issues/9287 to move the library to D3 in 1.0. It seems to support a broad range of chart types and might have greater benefits for us long term. We can sync up once we plan 1.0 to design the requirements there.

chirag-madlani commented:
> > 
> 
> @ShaileshParmar11 for this release we can go ahead with the scatter plot for the freshness. I opened a ticket #9287 to move the library to D3 in 1.0. It seems to support a broad range of chart types and might have greater benefits for us long term. We can sync up once we plan 1.0 to design the requirements there.

Hi @TeddyCr,
[Recharts](https://github.com/recharts/recharts) already build on top of [React](https://reactjs.org/) + [D3](https://github.com/d3)
I have worked with D3 as well and it's low-level lib which provides n number of combinations but at the same time complexity to understand and coding will be more. 
Recharts there to simplify the same. Recharts allow customisation in charts through which we can achieve whatever we are looking for. In case we don't achieve the same we can revisit the same. For now, I believe Recharts does satisfy our requirements

chirag-madlani commented on openmetadata-ui/src/main/resources/ui/src/components/Chart/OperationDateBarChart.tsx:
```diff
@@ -0,0 +1,110 @@
+        />
+        <CartesianGrid stroke={GRAPH_BACKGROUND_COLOR} />
+        <Tooltip formatter={tooltipFormatter} />
+        {information.map((info) => (
+          <Bar
+            barSize={1}
+            dataKey={info.dataKey}
+            fill={info.color}
+            hide={
+              activeKeys.length ? !activeKeys.includes(info.dataKey) : false
+            }
+            key={`${info.dataKey}-bar`}
+            name={info.title}
+            stackId="data"
+          />
+        ))}
+        {information.map((info) => (
+          <Scatter
+            dataKey={info.dataKey}
+            fill={info.color}
+            hide={
+              activeKeys.length ? !activeKeys.includes(info.dataKey) : false
+            }
+            key={`${info.dataKey}-scatter`}
+            name={info.title}
+          />
+        ))}
```
I feel now we can combine this to single loop but we can do that on next PR as well

ShaileshParmar11 replied:
Yes will do in next PR thanks
